### ISSUE TITLE
sql, opt: validating the unique constraint must be efficient for partitioned unique indexes

### ISSUE THREAD

**Issue by rytaft** [cockroachdb/cockroach#56201]
As discussed offline with @RaduBerinde:

When we create a partitioned unique index, we need to validate that we don't have duplicates on the unique column(s). If we write down a validation query for that today, it would likely end up being a hash aggregation on the full table, which is not acceptable. Instead, we would want to separate the scans between the regions so that each scan is ordered, and union them (keeping order) into a streaming aggregation.

To achieve this, we may be able to extend `SplitScanIntoUnionScans` (as described in https://github.com/cockroachdb/cockroach/issues/55156). Additionally, we currently only apply the rule when there's a limit, but we could add a corresponding rule for aggregations.

**Comment by rytaft** [comment]
Because of the need to fix #40797 as a prerequisite to this issue, I don't think this will make it into 21.1. It's unfortunate, but I think this is also not the highest priority item, since this validation query will only run when a new unique index is added to an existing RBR table. I'll try to get this in during the stability period, and if it seems low risk, it's a possible candidate for a backport.

**Comment by RaduBerinde** [comment]
Indeed, we would need UNION ALL to pass through orderings. I don't think that's a big change, but I also agree that this is not super high priority for the time being.

Just leaving a more detailed note about what this query would look like. I imagine it's something like `SELECT uniquecol GROUP BY uniquecol HAVING count(*) > 2` and we want that to be executed as
```
SELECT uniquecol FROM (
  (SELECT uniquecol WHERE region=region1 ORDER BY uniquecol) UNION ALL
  (SELECT uniquecol WHERE region=region2 ORDER BY uniquecol) UNION ALL
  (SELECT uniquecol WHERE region=region3 ORDER BY uniquecol)
)
GROUP BY uniquecol HAVING count(*) > 2
```
